java/ql/src/experimental/Security/CWE/CWE-352/JsonpInjectionLib.qll

```diff
@@ -0,0 +1,115 @@
+import java
+import DataFlow
+import JsonStringLib
+import semmle.code.java.security.XSS
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.DataFlow3
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.frameworks.spring.SpringController
+
+/**
+ * A method that is called to handle an HTTP GET request.
+ */
+abstract class RequestGetMethod extends Method {
+  RequestGetMethod() {
+    not exists(MethodAccess ma |
+      // Exclude apparent GET handlers that read a request entity, because this is the principle of JSONP.
```

Comment:
This is not an answer. WHAT is the principle of JSONP?

why do you believe this means JSONP injection cannot happen or is uninteresting?

or do you believe that is an indicator this is not a GET request?

Reply:
Maybe I should change the class name to JsonpRequestGetMethod.

Reply:
There is nothing wrong with the name. Please just tell me why do you believe that being able to reach `getInputStream` means this method is not an interesting case for JSONP injection?

Reply:
First of all, I am correcting that it is not that JSONP injection needs to be restricted, and JSONP itself is like this. JSONP exists to solve cross-domain data transmission. The src in the front-end <script src=""> inserts the request connection to obtain the data.  I can't imagine how to bring the body to the request.

Reply:
I don't know if I explain it this way or you can understand it.

Reply:
That does make sense, but because `GET` requests with a body are so unusual, and all subclasses of `RequestGetMethod` are specifically GET handler, why is this extra filter necessary? Have you found your query mistaking POST handlers for GET handlers, for example?

Reply:
The get request method can actually handle the request body. I think everyone knows this.  Why do I restrict it in RequestGetMethod? The purpose is to eliminate this situation.  Take the actual case above, the @RequestMapping annotation is used in SpringControllerMethod, and the method value is not specified in the annotation. At this time, the request method can be get/post/put, etc., then I can send a get request and there is a request body  , The program will process my request.

Reply:
Have you ever really seen a GET handler that accepts a body? It's such a strange thing to do, adding complication and cost to the query doesn't seem worth it unless you've seen this happen in the wild.

Reply:
1. Even if I don't want to believe that a get request handles the request body, I believe it exists.  If I don’t rule it out, there will be false positives when this happens, and I don’t want to see it.  2. The @RequestMapping annotation was used in the SpringController example I gave you, and the program processed the request body. The most important thing is that the program accepts user parameters as the JSONP method name to return. This is a true false positive.

Reply:
https://github.com/leiyi1152/shancai/blob/master/src/main/java/com/icloud/web/user/app/AppUserController.java#L50